File: contracts/tests/test_identified_multisig_wallet_with_daily_limit.py
# ethereum
from ethereum import tester as t
from ethereum.tester import keys, accounts
from preprocessor import PreProcessor
# standard libraries
from unittest import TestCase


class TestContract(TestCase):
    """
    run test with python -m unittest tests.test_identified_multisig_wallet_with_daily_limit
    """

    HOMESTEAD_BLOCK = 1150000
    TWENTY_FOUR_HOURS = 86400  # 24h

    def __init__(self, *args, **kwargs):
        super(TestContract, self).__init__(*args, **kwargs)
        self.s = t.state()
        self.s.block.number = self.HOMESTEAD_BLOCK
        t.gas_limit = 4712388
        self.pp = PreProcessor()

    def test(self):
        # Create wallet
        required_accounts = 2
        daily_limit = 1000
        wa_1 = 1
        wa_2 = 2
        wa_3 = 3
        id_1 = 123456789
        id_2 = 987654321
        id_3 = 192837465
        constructor_parameters = (
            [accounts[wa_1], accounts[wa_2], accounts[wa_3]],
            [id_1, id_2, id_3],
            required_accounts,
            daily_limit
        )
        gas = self.s.block.gas_used
        self.multisig_wallet = self.s.abi_contract(
            self.pp.process('IdentifiedMultiSigWalletWithDailyLimit.sol', contract_dir='solidity/', add_dev_code=True),
            language='solidity',
            constructor_parameters=constructor_parameters,
            contract_name="IdentifiedMultiSigWalletWithDailyLimit"
        )
        self.assertLess(self.s.block.gas_used - gas, 2000000)
        print "Deployment costs: {}".format(self.s.block.gas_used - gas)
        # Create ABIs
        multisig_abi = self.multisig_wallet.translator
        # Send money to wallet contract
        deposit = 10000
        self.s.send(keys[wa_1], self.multisig_wallet.address, deposit)
        self.assertEqual(self.s.block.get_balance(self.multisig_wallet.address), deposit)
        self.assertEqual(self.multisig_wallet.dailyLimit(), daily_limit)
        # Update daily limit
        daily_limit_updated = 2000
        update_daily_limit = multisig_abi.encode("changeDailyLimit", [daily_limit_updated])
        transaction_hash = self.multisig_wallet.submitTransaction(self.multisig_wallet.address, 0,
                                                                  update_daily_limit, 0, sender=keys[wa_1])
        self.multisig_wallet.confirmTransaction(transaction_hash, sender=keys[wa_2])
        self.assertEqual(self.multisig_wallet.dailyLimit(), daily_limit_updated)
        # Withdraw daily limit
        value = 1000
        wa_1_balance = self.s.block.get_balance(accounts[wa_1])
        nonce = self.multisig_wallet.getNonce(accounts[wa_1], value, "")
        self.multisig_wallet.submitTransaction(accounts[wa_1], value, "", nonce, sender=keys[wa_2])
        self.assertEqual(self.s.block.get_balance(self.multisig_wallet.address), deposit - value)
        self.assertEqual(self.s.block.get_balance(accounts[wa_1]), wa_1_balance + value)
        nonce = self.multisig_wallet.getNonce(accounts[wa_1], value, "")
        self.multisig_wallet.submitTransaction(accounts[wa_1], value, "", nonce, sender=keys[wa_2])
        self.assertEqual(self.s.block.get_balance(self.multisig_wallet.address), deposit - value*2)
        self.assertEqual(self.s.block.get_balance(accounts[wa_1]), wa_1_balance + value*2)
        # Third time fails, because daily limit was reached
        nonce = self.multisig_wallet.getNonce(accounts[wa_1], value, "")
        self.multisig_wallet.submitTransaction(accounts[wa_1], value, "", nonce, sender=keys[wa_2])
        self.assertEqual(self.s.block.get_balance(self.multisig_wallet.address), deposit - value*2)
        self.assertEqual(self.s.block.get_balance(accounts[wa_1]), wa_1_balance + value*2)
        # Let one day pass
        self.s.block.timestamp += self.TWENTY_FOUR_HOURS + 1
        # User wants to withdraw more than the daily limit. Withdraw is unsuccessful.
        value_2 = 3000
        nonce = self.multisig_wallet.getNonce(accounts[wa_1], value_2, "")
        self.multisig_wallet.submitTransaction(accounts[wa_1], value_2, "", nonce, sender=keys[wa_2])
        # Wallet and user balance remain the same.
        self.assertEqual(self.s.block.get_balance(self.multisig_wallet.address), deposit - value*2)
        self.assertEqual(self.s.block.get_balance(accounts[wa_1]), wa_1_balance + value*2)
        # Daily withdraw is possible again
        nonce = self.multisig_wallet.getNonce(accounts[wa_1], value, "")
        self.multisig_wallet.submitTransaction(accounts[wa_1], value, "", nonce, sender=keys[wa_2])
        # Wallet balance decreases and user balance increases.
        self.assertEqual(self.s.block.get_balance(self.multisig_wallet.address), deposit - value*3)
        self.assertEqual(self.s.block.get_balance(accounts[wa_1]), wa_1_balance + value*3)

File: contracts/tests/testIdentifiedMultiSigWalletWithDailyLimit.js
var async = require('async');
var chaiAsPromised = require('chai-as-promised');
var chai = require('chai');
var expect = chai.expect; // we are using the "expect" style of Chai
chai.use(chaiAsPromised);
var Web3 = require('web3');
var web3 = new Web3();
web3.setProvider(new web3.providers.HttpProvider('http://localhost:8545'));

var wa_1 = 1;
var wa_2 = 2;
var wa_3 = 3;
var wa_4 = 4;
var id_1 = 123456789;
var id_2 = 987654321;
var id_3 = 192837465;
var id_4 = 918273645;
var required_accounts = 2;
var accounts = web3.eth.accounts;

var identifiedMultiSigWallet;

var excludePending = false;
var includePending = true;
var excludeExecuted = false;
var includeExecuted = true;
async.series([
  function(callback) {
    describe('Deploy IdentifiedMultiSigWalletWithDailyLimit.', function() {
      it('Contract address should not be undefined', function(done) {
        var identifiedMultiSigWalletWithDailyLimitABI = require('../abi/IdentifiedMultiSigWalletWithDailyLimitABI.js')
        var identifiedMultiSigWalletWithDailyLimitBytecode = require('../bytecode/IdentifiedMultiSigWalletWithDailyLimitBytecode.js');
        var _owners = [accounts[wa_1], accounts[wa_2], accounts[wa_3]];
        var _ids = [id_1, id_2, id_3];
        var _required = required_accounts;
        var _dailyLimit = web3.toWei(10);
        var identifiedMultiSigWalletWithDailyLimitContract = web3.eth.contract(identifiedMultiSigWalletWithDailyLimitABI);
        identifiedMultiSigWalletWithDailyLimit = identifiedMultiSigWalletWithDailyLimitContract.new(
         _owners,
         _ids,
         _required,
         _dailyLimit,
         {
           from: accounts[wa_1],
           data: identifiedMultiSigWalletWithDailyLimitBytecode,
           gas: '4700000'
         },
         function (e, contract){
            if (typeof contract.address !== 'undefined') {
              expect(parseInt(identifiedMultiSigWalletWithDailyLimit.dailyLimit().toString())).to.equal(parseInt(web3.toWei(10).toString()));
              done();
              callback();
            }
         });
      });
    });
  },
  function(callback) {
    describe('Changing daily limit:', function () {
      it('Daily limit should be changed', function(done) {
        let nonce = identifiedMultiSigWalletWithDailyLimit.getNonce(web3.toWei(5));
        let transactionData = identifiedMultiSigWalletWithDailyLimit.changeDailyLimit.getData(web3.toWei(5));
        identifiedMultiSigWalletWithDailyLimit.submitTransaction(identifiedMultiSigWalletWithDailyLimit.address, 0, transactionData, nonce, {
          from: accounts[wa_1],
          to: identifiedMultiSigWalletWithDailyLimit.address,
          value: 0,
          gas: '3000000'
        }, function(e, txHash) {
          let hash = identifiedMultiSigWalletWithDailyLimit.getTransactionHashes(0, identifiedMultiSigWalletWithDailyLimit.getTransactionCount(includePending, includeExecuted), includePending, excludeExecuted)[0];
          identifiedMultiSigWalletWithDailyLimit.confirmTransaction(hash,{
            from: accounts[wa_2],
            to: identifiedMultiSigWalletWithDailyLimit.address,
            gas: '3000000'
          }, function(e, res) {
            expect(parseInt(identifiedMultiSigWalletWithDailyLimit.dailyLimit().toString())).to.equal(parseInt(web3.toWei(5).toString()));
            done();
            callback();
          });
        });
      });
    });
  }, function(callback) {
    describe('Sending Money with daily limit:', function() {
      it('Money should be sent succesfully to the wallet', function(done) {
        web3.eth.sendTransaction({
          from: accounts[wa_1],
          to: identifiedMultiSigWalletWithDailyLimit.address,
          value: web3.toWei(10),
          gas: '3000000'
        });
        expect(parseInt(web3.eth.getBalance(identifiedMultiSigWalletWithDailyLimit.address).toString())).to.equal(parseInt(web3.toWei(10).toString()));
        done();
      });
      it('Money should be sent succesfully from the wallet', function(done) {
        let nonce = identifiedMultiSigWalletWithDailyLimit.getNonce(accounts[0], web3.toWei(5), "");
        identifiedMultiSigWalletWithDailyLimit.submitTransaction(accounts[0], web3.toWei(5), "", nonce, {
          from: accounts[wa_1],
          to: identifiedMultiSigWalletWithDailyLimit.address,
          value: 0,
          gas: '3000000'
        }, function(e, txHash) {
          let hash = identifiedMultiSigWalletWithDailyLimit.getTransactionHashes(0, identifiedMultiSigWalletWithDailyLimit.getTransactionCount(includePending, includeExecuted), includePending, excludeExecuted)[0];
          identifiedMultiSigWalletWithDailyLimit.confirmTransaction(hash,{
            from: accounts[wa_2],
            to: identifiedMultiSigWalletWithDailyLimit.address,
            gas: '3000000'
          }, function(e, res) {
            expect(parseInt(web3.eth.getBalance(identifiedMultiSigWalletWithDailyLimit.address).toString())).to.equal(parseInt(web3.toWei(5).toString()));
            done();
          });
        });
      });
      it('Money should not be sent succesfully from the wallet when the daily limit is reached', function() {
        let nonce = identifiedMultiSigWalletWithDailyLimit.getNonce(accounts[0], web3.toWei(5), "");
        identifiedMultiSigWalletWithDailyLimit.submitTransaction(accounts[0], web3.toWei(5), "", nonce, {
          from: accounts[wa_1],
          to: identifiedMultiSigWalletWithDailyLimit.address,
          value: 0,
          gas: '3000000'
        }, function(e, txHash) {
          let hash = identifiedMultiSigWalletWithDailyLimit.getTransactionHashes(0, identifiedMultiSigWalletWithDailyLimit.getTransactionCount(includePending, includeExecuted), includePending, excludeExecuted)[0];
          identifiedMultiSigWalletWithDailyLimit.confirmTransaction(hash,{
            from: accounts[wa_2],
            to: identifiedMultiSigWalletWithDailyLimit.address,
            gas: '3000000'
          }, function(e, res) {
            expect(parseInt(web3.eth.getBalance(identifiedMultiSigWalletWithDailyLimit.address).toString())).to.equal(5);
            callback();
          });
        });
      });
    });
  },
]);
